src/book/summary.rs

```diff
@@ -0,0 +1,716 @@
+                let link = Link::new(text, location.as_ref());
+
+                debug!("[*] Found a prefix chapter: {:?}", link.name);
+                self.summary.prefix_chapters.push(SummaryItem::Link(link));
+            },
+            Event::End(Tag::Rule) => {
+                debug!("[*] Found a prefix chapter separator");
+                self.summary.prefix_chapters.push(SummaryItem::Separator);
+            },
+            Event::Start(Tag::List(_)) => {
+                debug!("[*] Changing from prefix chapters to numbered chapters");
+                self.state = State::NumberedChapters(0);
+            },
+
+            other => {
+                trace!("[*] Skipping unexpected token in summary: {:?}", other);
+            },
+        }
+
+        Ok(())
+    }
+
+    /// Parse the numbered chapters.
+    ///
+    /// If the event is the start of a list item, consume the entire item and
+    /// add a new link to the summary with `push_numbered_section`.
+    ///
+    /// If the event is the start of a new list, bump the nesting level.
+    ///
+    /// If the event is the end of a list, decrement the nesting level. When
+    /// the nesting level would go negative, we've finished the numbered
+    /// section and need to parse the suffix section.
+    ///
+    /// Otherwise, ignore the event.
+    fn step_numbered(&mut self, event: Event) -> Result<()> {
+        match event {
+            Event::Start(Tag::Item) => {
+                let it = self.parse_item().chain_err(
+                    || "List items should only contain links",
+                )?;
+
+                debug!("[*] Found a chapter: {:?} ({})", it.name, it.location.display());
+                let section_number = self.push_numbered_section(SummaryItem::Link(it));
+                trace!("[*] Section number is {}", section_number);
+            },
+            Event::Start(Tag::List(_)) => {
+                if let State::NumberedChapters(n) = self.state {
+                    self.state = State::NumberedChapters(n + 1);
+                    trace!("[*] Nesting level increased to {}", n + 1);
+                }
+            },
+            Event::End(Tag::List(_)) => {
+                if let State::NumberedChapters(n) = self.state {
+                    if n == 0 {
+                        trace!("[*] Finished parsing the numbered chapters");
+                        self.state = State::SuffixChapters;
+                    } else {
+                        trace!("[*] Nesting level decreased to {}", n - 1);
+                        self.state = State::NumberedChapters(n - 1);
+                    }
+                }
+            },
+            other => {
+                trace!("[*] skipping unexpected token: {:?}", other);
+            },
+        }
+
+        Ok(())
+    }
+
+    fn step_suffix(&mut self, event: Event<'a>) -> Result<()> {
+        // FIXME: This has been copy/pasted from step_prefix. make DRY.
```

Comment:
how about passing state `State::PrefixChapters`|`State::SuffixChapters` to `step_prefix` (renaming it to something like `step_affix`) and remove the copy pasted code?